feathr_project/test/test_pyduf_preprocessing_e2e.py

```diff
@@ -0,0 +1,456 @@
+import os
+from datetime import datetime, timedelta
+from pathlib import Path
+from pyspark.sql.functions import col,sum,avg,max
+from feathr.job_utils import get_result_df
+from feathr.anchor import FeatureAnchor
+from feathr.dtype import STRING, BOOLEAN, FLOAT, INT32, ValueType
+from feathr.feature import Feature
+from feathr.query_feature_list import FeatureQuery
+from feathr.settings import ObservationSettings
+from feathr.source import INPUT_CONTEXT, HdfsSource
+from feathr.typed_key import TypedKey
+from feathr.transformation import WindowAggTransformation
+from pyspark.sql import SparkSession, DataFrame
+from test_fixture import basic_test_setup
+from feathr import (BackfillTime, MaterializationSettings)
+from feathr import RedisSink
+from test_fixture import snowflake_test_setup
+
+def trip_distance_preprocessing(df: DataFrame):
+    df = df.withColumn("trip_distance", df.trip_distance.cast('double') - 90000)
+    df = df.withColumn("fare_amount", df.fare_amount.cast('double') - 90000)
+
+    return df
+
+def add_new_dropoff_and_fare_amount_column(df: DataFrame):
+    df = df.withColumn("new_lpep_dropoff_datetime", col("lpep_dropoff_datetime"))
+    df = df.withColumn("new_fare_amount", col("fare_amount") + 1000000)
+    return df
+
+def add_new_fare_amount(df: DataFrame) -> DataFrame:
+    df = df.withColumn("fare_amount_new", col("fare_amount") + 8000000)
+
+    return df
+
+def add_new_surcharge_amount_and_pickup_column(df: DataFrame) -> DataFrame:
+    df = df.withColumn("new_improvement_surcharge", col("improvement_surcharge") + 1000000)
+    df = df.withColumn("new_tip_amount", col("tip_amount") + 1000000)
+    df = df.withColumn("new_lpep_pickup_datetime", col("lpep_pickup_datetime"))
+
+    return df
+
+def add_old_lpep_dropoff_datetime(df: DataFrame) -> DataFrame:
+    df = df.withColumn("old_lpep_dropoff_datetime", col("lpep_dropoff_datetime"))
+
+    return df
+
+def feathr_udf_day_calc(df: DataFrame) -> DataFrame:
+    df = df.withColumn("f_day_of_week", dayofweek("lpep_dropoff_datetime"))
```

Comment:
Where is 'dayofweek' from?

Reply:
it's from pyspark udf which we imports in driver code.